onnxruntime/core/optimizer/graph_transformer_utils.cc

```diff
@@ -108,14 +98,22 @@ std::vector<std::unique_ptr<GraphTransformer>> GenerateTransformers(TransformerL
       break;
   }
 
-  // If the rule-based transformer is not empty, it should be included in the custom transformer list below.
-  if (non_empty_rule_transformer) {
-    transformers_and_rules_to_enable->push_back(transformer_utils::GenerateRuleBasedTransformerName(level));
-  }
-  if (transformers_and_rules_to_enable != nullptr && !transformers_and_rules_to_enable->empty()) {
-    // pick custom transformers enabled for this session
+  // if the custom list to enable transformers\rules is empty then return the default generated transformers and rules
+  // otherwise generate a filtered list based on the provided custom list.
+  if (transformers_and_rules_to_enable.empty()) {
+    if (rule_transformer != nullptr) {
```

Comment:
The way GenerateRuleBasedGraphTransformer is implemented, we need to check whether there are any rules rather than if it is nullptr. I don't think it can be nullptr.

Reply:
In the current implementation when custom list is empty it is not really creating any object of RuleBasedTransformer.  
However to make it clear I have already changed the code to explicitly return nullptr when custom list is empty.  My changes after merge master were not added to the commit so it did not reflect here. I have pushed the latest changes now

I have also added a test to specifically test only and all the expected transformers in the custom list are returned.

Reply:
Cool, works now.
Nit: for the nullptr check, you can simply do `if (rule_transformer)` here and below.

Reply:
I know I prefer this as it is very clear...